rules/do-you-use-strongly-typed-ids/rule.md

```diff
@@ -0,0 +1,138 @@
+
+## How do we configure EF to work with Strongly Typed IDs
+
+To enable the correct serialization of Strongly Type IDs we need to add some configuration code to our EF Entities so that EF knows how to convert between the Primitive type and the Strongly Typed ID values.
+
+Here is one way to specify the configuration:
+
+```cs
+public class PersonConfiguration : IEntityTypeConfiguration<Person>
+{
+    public void Configure(EntityTypeBuilder<Person> builder)
+    {
+        // Primary Key
+        builder.HasKey(e => e.Id);
+
+        // Configuration for converting the `id` property To that will be stored in the database and From the primitive value (e.g. GUID, int, string) back into a strongly typed Id:
+        builder.Property(e => e.Id)
+            .HasConversion(
+                // from Strong ID to Primitive value
+                id => id.Value,
+                // from Primitive value to Strong ID
+                guidValue => new PersonId(guidValue))
+            // enable the primitive value to be auto-generated when adding (SaveChanges) new entities to the database (only works from EF Core 7 onwards)
+            .ValueGeneratedOnAdd();
+    }
+}
+```
+
+Another slightly cleaner looking approach could be to use a `ValueConverter`:
+
+```cs
+public class PersonConfiguration : IEntityTypeConfiguration<Person>
+{
+    public void Configure(EntityTypeBuilder<Person> builder)
+    {
+        var converter = new ValueConverter<PersonId, Guid>(
+            id => id.Value, 
+            guidValue => new PersonId(guid));
+
+        builder.HasKey(e => e.Id);
+
+        builder.Property(e => e.Id)
+            .HasConversion(converter)
+            .ValueGeneratedOnAdd();
+    }
+}
+```
+
+Check out the [SSW Clean Architecture template](https://github.com/SSWConsulting/SSW.CleanArchitecture/blob/main/src/Domain/TodoItems/TodoItem.cs) for a few examples of Strongly Typed IDs in action.
+
+## Performance considerations
+
+As mentioned earlier, we could use `record` or `record struct` to succinctly describe our strongly typed IDs. When using `record` we get equality checks automatically generated for us by the compiler.
+
+However, when using `record` there is a performance penalty we pay as pointed out by Dan Patrascu on his CodeWrinkles YouTube video [Stop Using Records As Strongly Typed IDs!](https://www.youtube.com/watch?v=dJxVj6390hk).
+
+Dan suggests to use `struct` or `class` instead because generating 10,000 strongly typed Ids is 4x faster compared to using `record`.
+
+However, `class` and `struct` are more verbose and require that we implement the `Equals` and `GetHashCode` overrides manually!
+
+Dan unfortunately missed out a benchmark for `record struct` which turns out to give us the best of both worlds - performance and succinct one liner declarations without manually implementing equality checks.
+
+So the ultimate programming construct for Strongly Typed IDs is `readonly record struct`.
```

Comment:
Why `readonly record struct` over `record struct`?  Do you have a benchmark result for that also?

Reply:
`readonly` only helps with immutability in some cases (not all). The compiler might throw up some compile errors when it can detect internal state being modified instead of just being read.

Check out this page for lots of examples: https://steven-giesel.com/blogPost/2a52cd8d-b3b2-42e4-87e8-d6dc14147ddb

I link to this in the benchmark app too.